istioctl/cmd/injector-list.go

```diff
@@ -0,0 +1,328 @@
+		Use:     "list",
+		Short:   "List sidecar injector and sidecar versions",
+		Long:    `List sidecar injector and sidecar versions`,
+		Example: `  istioctl experimental injector list`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client, err := kubeClientWithRevision(kubeconfig, configContext, opts.Revision)
+			if err != nil {
+				return fmt.Errorf("failed to create k8s client: %v", err)
+			}
+			nslist, err := getNamespaces(client)
+			if err != nil {
+				return err
+			}
+			hooks, err := getWebhooks(client)
+			if err != nil {
+				return err
+			}
+			pods, err := getPods(client)
+			if err != nil {
+				return err
+			}
+			err = printNS(cmd.OutOrStdout(), nslist, hooks, pods)
+			if err != nil {
+				return err
+			}
+			cmd.Println()
+			injectedImages, err := getInjectedImages(client)
+			if err != nil {
+				return err
+			}
+			return printHooks(cmd.OutOrStdout(), nslist, hooks, injectedImages)
+		},
+	}
+
+	return cmd
+}
+
+func getNamespaces(client kube.ExtendedClient) ([]v1.Namespace, error) {
+	nslist, err := client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return []v1.Namespace{}, err
+	}
+	return nslist.Items, nil
+}
+
+func printNS(writer io.Writer, namespaces []v1.Namespace, hooks []admit_v1.MutatingWebhookConfiguration,
+	allPods map[resource.Namespace][]v1.Pod) error {
+
+	sort.Slice(namespaces, func(i, j int) bool {
+		return namespaces[i].Name < namespaces[j].Name
+	})
+
+	outputCount := 0
+
+	w := new(tabwriter.Writer).Init(writer, 0, 8, 1, ' ', 0)
+	for _, namespace := range namespaces {
+		if hideFromOutput(resource.Namespace(namespace.Name)) {
+			continue
+		}
+
+		revision := getInjectedRevision(&namespace, hooks)
+		podCount := podCountByRevision(allPods[resource.Namespace(namespace.Name)], revision)
+		if len(podCount) == 0 {
+			// This namespace has no pods, but we wish to display it if new pods will be auto-injected
+			if revision != "" {
+				podCount[revision] = revisionCount{}
+			}
+		}
+		for injectedRevision, count := range podCount {
+			if outputCount == 0 {
+				fmt.Fprintln(w, "NAMESPACE\tISTIO-REVISION\tPOD-REVISIONS")
+			}
+			outputCount++
+
+			fmt.Fprintf(w, "%s\t%s\t%s\n", namespace.Name, revision, renderCounts(injectedRevision, count))
+		}
+	}
+	if outputCount == 0 {
+		fmt.Fprintf(writer, "No Istio injected namespaces present.\n")
+	}
+
+	return w.Flush()
+}
+
+func getWebhooks(client kube.ExtendedClient) ([]admit_v1.MutatingWebhookConfiguration, error) {
+	hooks, err := client.AdmissionregistrationV1().MutatingWebhookConfigurations().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return []admit_v1.MutatingWebhookConfiguration{}, err
+	}
+	return hooks.Items, nil
+}
+
+func printHooks(writer io.Writer, namespaces []v1.Namespace, hooks []admit_v1.MutatingWebhookConfiguration, injectedImages map[string]string) error {
+	if len(hooks) == 0 {
+		_, err := fmt.Fprintf(writer, "No Istio injection hooks present.\n")
```

Comment:
super nit: It feels a bit weird to call this an error and then not check it before calling printHooks above. It doesn't feel like an error condition to try and print something that's not there--otherwise print("") should throw an error.

I'd rather just return early, and then print this to stderr (or ErrOrStdErr()) instead of returning it as an error.

Reply:
I will change the code to ignore this.  This detects I/O error, but it seems pointless to check for error, as if there is no where to write output there is no where to complain about being unable to write output.